### ISSUE TITLE
Support for non-http backends

### ISSUE THREAD

**Issue by mortenab** [containous/traefik-extra-service-fabric#41]
Hi,

We are exposing gRPC services through Traefik with the h2c protocol on the backends. Because of the logic in `getValidInstances` shown below, service instances that do not have an http endpoint are considered invalid:

https://github.com/containous/traefik-extra-service-fabric/blob/6e90a9eef2ac9d320e55d6e994d169673a8d8b0f/servicefabric.go#L202-L215

As a consequence, the gRPC service instances are not included when the configuration template is evaluated.

We have managed to work around this issue by first adding an additional "dummy" http endpoint to the services and then writing a custom configuration template that selects the "h2c" endpoint using the `getNamedEndpoint` template function.

Our workaround seems a bit fragile and it would be useful with a first order support for custom protocols - for example by using the `traefik.protocol` label on the service.

**Comment by jjcollinge** [comment]
Agree, I think the best approach is to have a func `hasRoutableEndpoint` that validates the service's protocol matches a supported list of Traefik protocols (http://, https://, h2c:// and check for other HTTP2 variants). I've called it `...Endpoint` as it will allow us to extend functionality later to validate other parts of the URI schema should we need to.  I think we should handle this without the need for an additional label.

**Comment by mortenab** [comment]
Hi,

Thanks for looking into this! Sounds like a good solution - I suggested the `traefik.protocol` label as it seems to be preferred by other providers. 

As far as I can tell, Service Fabric does not support specifying `UriScheme`s other than `tcp`, `http` and `https` so the custom scheme won't actually be specified anywhere in the service manifest. We have worked around this limitation by creating our own `ICommunicationListener` that returns a URL on the form `h2c://<server>:<port>`  in the `OpenAsync()` method. This is the endpoint URL that the Traefik provider ends up receiving.

**Comment by jjcollinge** [comment]
I know having to override SF URI behaviour isn't ideal - however, that allows us to keep a single source of truth. Introducing a new label would be a potential point of conflict between our provider and the cluster configuration. Implementing an `ICommunicationListen` to return a custom endpoint is a fairly common pattern in SF and as long as we point to docs on how to do it I think it should be ok to do this without a new label. Happy to discuss further, this isn't a particular pain point I have to live with so @mortenab your views are key.

**Comment by mortenab** [comment]
I actually don't mind overriding the URI in the  `ICommunicationListener` as we would have implemented the `ICommunicationListener` anyway. The main issues for me are:
1. We are required to override the built-in template when using the h2c protocol
2. Inside the template we need to hard-code the name of the endpoint returned by the `ICommunicationListener`.

As far as I can see, the solution proposed by @jjcollinge will solve both problems 👍

**Comment by lawrencegripper** [comment]
I agree, sorry for complicating matters! @mortenab are you happy to submit a PR to add this functionality?

**Comment by mortenab** [comment]
Yes - I can submit a PR. I do not have the bandwidth the coming weeks, but I will put it on the backlog.
